### ISSUE TITLE
Strange inspection warnings on `assert()` after `while(true)`

### ISSUE THREAD

**Issue by donquixote** [vimeo/psalm#7262]
Code with inspection warnings:

https://psalm.dev/r/829fbbf65b

```php
function test(bool $b): void {
    while (TRUE) {
        assert($b);  // Type true for TRUE is never falsy
    }
}

function test2(bool $b, bool $c): void {
    while (true) {
        \assert($b);
        while (true) {
            \assert($c);  // Docblock-defined type non-empty-mixed for true is never falsy
        }
    }
}
```

Warrnings:
"Type true for TRUE is never falsy"
"Docblock-defined type non-empty-mixed for true is never falsy"

I see the following problems:
- These warnings seem to only occur in this combination, not for the specific parts as standalone. E.g. `while (TRUE)` might be suspicious in itself, but it does not trigger the inspection warning.
- If the warnings refer to the `while (true)` / `while (TRUE)`, it is reported on the wrong line.

Btw, you can add counter and exit conditions to the loops, they won't change a thing.
I just wanted to keep it as minimal as possible.

**Comment by orklah** [comment]
Psalm must check the loop a first time and assert that the variable is indeed true, then when it tries to do a second loop to see if the context changes, it encounters an assert with an always true value. 

I never really looked how loops work exactly, so I don't have a lot of details on what happens but I've seen this pattern a few times

**Comment by donquixote** [comment]
This explanation would make sense in the simplified example, because the value of `$b` or `$c` does not change between iterations.

However, we can make an advanced example where the condition should no longer be an invariant.
I am also inserting `print 'helllo';` so that it is not a "next line" problem.

https://psalm.dev/r/e2363c56bc

```php
function test(int $b): void {
    $i = 0;
    while (TRUE) {
        print 'hello';
        \assert($b > $i);  // Type true for TRUE is never falsy
    }
    ++$i;
    if ($i > 3) break;
}

function test2(int $b, int $c): void {
    $i = 0;
    while (true) {
        print 'hello';
        \assert($b > $i);
        $j = 0;
        while (true) {
            print 'hello';
            \assert($c > $j);  // Docblock-defined type non-empty-mixed for true is never falsy
        	++$j;
            if ($j > 3) break;
        }
        ++$i;
        if ($i > 3) break;
    }
}
```

**Comment by donquixote** [comment]
Btw, for repeated `assert(true);` it complains that "true has already been asserted".
https://psalm.dev/r/42ddbba4bd
Which is treating `true` as if it was a variable, not a guaranteed const expression.
It would make sense for user constants which could depend on other parts of code, or even which file was included.
But for `true`, it would make more sense to complain "assertion is always true" and be done with it.

**Comment by donquixote** [comment]
Another version, where now the asserted expression is completely unknown.
https://psalm.dev/r/39a505130b

**Comment by orklah** [comment]
I reverted a previous PR that was messing with the examples here. Can you check if it works as expected now?

**Comment by donquixote** [comment]
Indeed! With `vimeo/psalm:dev-master` version, the inspection warning no longer occurs.
Also, the problem occured only since a recent update.
Thanks!
Are you going to make a tagged release soon?

**Comment by orklah** [comment]
Done :)
